moveit_py/src/moveit/moveit_core/kinematic_constraints/utils.cpp

```diff
@@ -0,0 +1,158 @@
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Peter David Fagan */
+
+#include "utils.h"
+#include <rclcpp/rclcpp.hpp>
+#include <moveit_py/moveit_py_utils/ros_msg_typecasters.h>
+#include <moveit/kinematic_constraints/utils.h>
+
+namespace moveit_py
+{
+namespace bind_kinematic_constraints
+{
+moveit_msgs::msg::Constraints construct_link_constraint(const std::string& link_name, const std::string& source_frame,
+                                                        std::optional<std::vector<double>> cartesian_position,
+                                                        std::optional<double> cartesian_position_tolerance,
+                                                        std::optional<std::vector<double>> orientation,
+                                                        std::optional<double> orientation_tolerance)
+{
+  // check that link cartesian and/or orientation constraints are specified
+  if (!cartesian_position && !orientation)
+  {
+    throw std::invalid_argument("No link cartesian or orientation constraints specified");
+  }
+
+  moveit_msgs::msg::Constraints constraints_cpp;
+
+  // merge constraints if necessary
+  if (cartesian_position && orientation)
+  {
+    // define point stamped message
+    geometry_msgs::msg::PointStamped point;
+    point.header.frame_id = source_frame;
+    point.point.x = cartesian_position.value()[0];
+    point.point.y = cartesian_position.value()[1];
+    point.point.z = cartesian_position.value()[2];
+
+    moveit_msgs::msg::Constraints position_constraints =
+        kinematic_constraints::constructGoalConstraints(link_name, point, cartesian_position_tolerance.value());
+
+    // define quaternion message
+    geometry_msgs::msg::QuaternionStamped quaternion;
+    quaternion.header.frame_id = source_frame;
+    quaternion.quaternion.x = orientation.value()[0];
+    quaternion.quaternion.y = orientation.value()[1];
+    quaternion.quaternion.z = orientation.value()[2];
```

Comment:
**IMPORTANT**: Please take a look at this -- is this a severe bug?

Quaternions have 4 values, and there should also be a `w` value!
https://docs.ros2.org/latest/api/geometry_msgs/msg/Quaternion.html

True in other places as well

Reply:
This is a bug it should be a PointStamped message.

Reply:
This is a bug.